README.md
# jar-graph

[![Build Status](https://travis-ci.org/krlvi/jar-graph.svg?branch=master)](https://travis-ci.org/krlvi/jar-graph)

Graph analysis of the internal dependencies within a java archive file (JAR) using the [gephi toolkit](https://github.com/gephi/gephi-toolkit).

## Installation

Required is Java 8 or later. Download [the latest release](https://github.com/krlvi/jar-graph/releases/latest) and extract the archive.

## Usage
```
usage: ./jar-graph.sh -f <jar_file> -p <package> [-t time] [-nh] [-o out_file] [-w dir] [-r repulsion] [-s font] [-m min] [-x max]
  -t  time      specify simulation time in seconds
  -n            no graph will be generated
  -h            display help
  -o  outfile   output file outfile
  -w  workdir   working directory
  -r  repulsion node repulsion strength value
  -s  size      font size
  -m  min       minimum node size
  -x  max       maximum node size
```

## Examples



## License

Copyright © 2018 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-09-23
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-09-23
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/jar-graph/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/jar-graph/compare/0.1.0...0.1.1


test/jar_graph/core_test.clj
(ns jar-graph.core-test
  (:require [clojure.test :refer :all]
            [jar-graph.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/jar_graph/core.clj
(ns jar-graph.core
  (:require [clojure.tools.cli :refer [parse-opts]]
            [jar-graph.graph :as graph]
            [clojure.pprint :refer [pprint]])
  (:gen-class))

(defn exit [msg]
  (println msg)
  (System/exit 0))

(def cli-options
  [["-d" "--dot-file DOTFILE" "Path to the dot graph file to analyze"
    :required "Provide dot file to analyze"]
   ["-s" "--skip-graph" "Skip drawing graph PDF"]
   ["-o" "--out-graph OUT" "File name of the output graph PDF"]
   ["-t" "--simulation-time TIME" "Duration of clustering simulation in seconds"
    :default 10
    :parse-fn #(Integer/parseInt %)]
   ["-r" "--repulsion-strength REPULSION" "Node repulsion strength"
    :default 800
    :parse-fn #(Integer/parseInt %)]
   ["-n" "--min-size MINSIZE" "Size of the least connected nodes"
    :default 6
    :parse-fn #(Integer/parseInt %)]
   ["-x" "--max-size MAXSIZE" "Size of the most connected nodes"
    :default 40
    :parse-fn #(Integer/parseInt %)]
   ["-l" "--label-size LABELSIZE" "Label font size"
    :default 2
    :parse-fn #(Integer/parseInt %)]
   ["-h" "--help" "Prints this help message"]])

(defn -main [& args]
  (let [{:keys [options errors summary]} (clojure.tools.cli/parse-opts args cli-options)]
    (cond
      (:help options) (exit summary)
      errors (exit errors)
      (:dot-file options) (clojure.pprint/pprint (graph/analyze options))
      :else (exit summary))))


src/jar_graph/graph.clj
(ns jar-graph.graph
  (:require [com.stuartsierra.frequencies :as freq])
  (:import [java.util.concurrent TimeUnit]
           [org.openide.util Lookup]
           [org.gephi.project.api ProjectController]
           [org.gephi.io.importer.api ImportController EdgeDirectionDefault]
           [org.gephi.io.processor.plugin DefaultProcessor]
           [org.gephi.io.exporter.api ExportController]
           [org.gephi.graph.api GraphController]
           [org.gephi.layout.plugin AutoLayout]
           [org.gephi.layout.plugin.forceAtlas ForceAtlasLayout]
           [org.gephi.appearance.api AppearanceController AppearanceModel$GraphFunction]
           [org.gephi.appearance.plugin RankingNodeSizeTransformer PartitionElementColorTransformer]
           [org.gephi.statistics.plugin Modularity]
           [org.gephi.layout.plugin.noverlap NoverlapLayout]
           [org.gephi.preview.api PreviewController PreviewProperty]
           [org.gephi.layout.plugin.labelAdjust LabelAdjust]
           [org.gephi.preview.types DependantOriginalColor]
           [org.gephi.appearance.plugin.palette PaletteManager])
  (:gen-class))

(defn- look-up [class]
  (.. (Lookup/getDefault) (lookup class)))

(defn- graph-model []
  (.. (look-up GraphController) (getGraphModel)))

(defn- workspace []
  (let [pc (look-up ProjectController)
        ws (.. pc (getCurrentWorkspace))]
    (if ws
      ws
      (do
        (.. pc (newProject))
        (.. pc (getCurrentWorkspace))))))

(defn- get-container [file-path import-controller]
  (let [file (java.io.File. file-path)
        container (.. import-controller (importFile file))]
    (.. container (getLoader) (setEdgeDefault EdgeDirectionDefault/DIRECTED))
    (.. container (getLoader) (setAllowAutoNode false))
    container))

(defn- import-to-workspace [file-path]
  (let [import-controller (look-up ImportController)
        container (get-container file-path import-controller)
        processor (new DefaultProcessor)]
    (.. import-controller (process container processor (workspace)))))

(defn- gen-stats []
  (let [graph (.. (graph-model) (getDirectedGraph))
        nodes (.. graph (getNodes) (toCollection))
        in-degree (map (fn [n] (.. graph (getInDegree n))) nodes)
        out-degree (map (fn [n] (.. graph (getOutDegree n))) nodes)
        degree (map (fn [n] (.. graph (getDegree n))) nodes)]
    {:in-degree (freq/stats (frequencies in-degree))
     :out-degree (freq/stats (frequencies out-degree))
     :degree (freq/stats (frequencies degree))}))

(defn- apply-size-by-degree [min-size max-size]
  (let [appearance-controller (look-up AppearanceController)
        appearance-model (.. appearance-controller (getModel))
        graph (.. (graph-model) (getDirectedGraph))
        degree-ranking (.. appearance-model (getNodeFunction graph AppearanceModel$GraphFunction/NODE_INDEGREE,
                                                             RankingNodeSizeTransformer))]
    (.. degree-ranking (getTransformer) (setMinSize min-size))
    (.. degree-ranking (getTransformer) (setMaxSize max-size))
    (.. appearance-controller (transform degree-ranking))))

(defn- apply-color-by-partition []
  (.. (new Modularity) (execute (graph-model)))
  (let [appearance-controller (look-up AppearanceController)
        mod-column (.. (graph-model) (getNodeTable) (getColumn Modularity/MODULARITY_CLASS))
        node-func (.. appearance-controller (getModel) (getNodeFunction (.. (graph-model) (getDirectedGraph))
                                                                        mod-column PartitionElementColorTransformer))
        partition (.. node-func (getPartition))
        palette (.. (PaletteManager/getInstance) (randomPalette (.. partition (size))))
        ]
    (.. partition (setColors (.. palette (getColors))))
    (.. appearance-controller (transform node-func))))

(defn- apply-labels [font-size dark-mode]
  (let [props (.. (.. (look-up PreviewController) (getModel)) (getProperties))]
    (.. props (putValue PreviewProperty/DIRECTED Boolean/TRUE))
    (.. props (putValue PreviewProperty/EDGE_CURVED Boolean/FALSE))
    (.. props (putValue PreviewProperty/SHOW_NODE_LABELS Boolean/TRUE))
    (.. props (putValue PreviewProperty/NODE_LABEL_PROPORTIONAL_SIZE Boolean/FALSE))
    (.. props (putValue PreviewProperty/EDGE_THICKNESS 0.3))
    (.. props (putValue PreviewProperty/EDGE_OPACITY 85))
    (if dark-mode
      (do
        (.. props (putValue PreviewProperty/BACKGROUND_COLOR java.awt.Color/BLACK))
        (.. props (putValue PreviewProperty/NODE_LABEL_COLOR (new DependantOriginalColor java.awt.Color/WHITE)))))
    (.. props (putValue PreviewProperty/NODE_LABEL_FONT (new java.awt.Font nil 0 font-size)))
    (.. (look-up PreviewController) (refreshPreview))))

(defn- apply-layout [time-sec repulsion-strength]
  (let [auto-layout (new AutoLayout time-sec TimeUnit/SECONDS)
        no-overlap (new NoverlapLayout nil)
        force-atlas (new ForceAtlasLayout nil)
        label-adjust (new LabelAdjust nil)
        prop-adjust-by-size (AutoLayout/createDynamicProperty "forceAtlas.adjustSizes.name" Boolean/TRUE 0.0)
        prop-repulsion (AutoLayout/createDynamicProperty "forceAtlas.repulsionStrength.name" (new Double (double repulsion-strength)) 0.0)]
    (.. auto-layout (setGraphModel (graph-model)))
    (.. auto-layout (addLayout force-atlas 0.90 (into-array [prop-adjust-by-size prop-repulsion])))
    (.. auto-layout (addLayout no-overlap 0.05))
    (.. auto-layout (addLayout label-adjust 0.05))
    (.. auto-layout (execute))))

(defn- export-to-pdf [out-file]
  (let [export-controller (look-up ExportController)
        pdf-exporter (.. export-controller (getExporter "pdf"))
        baos (new java.io.ByteArrayOutputStream)
        out-file (new java.io.FileOutputStream out-file)]
    (.. pdf-exporter (setPageSize com.itextpdf.text.PageSize/A0))
    (.. pdf-exporter (setWorkspace (workspace)))
    (.. export-controller (exportStream baos pdf-exporter))
    (.. out-file (write (.. baos (toByteArray))))
    (.. baos (close))
    (.. out-file (close))))

(defn analyze [options]
  (let [{:keys [dot-file skip-graph out-graph simulation-time repulsion-strength min-size max-size label-size]} options]
    (import-to-workspace dot-file)
    (if-not skip-graph
      (do
        (apply-size-by-degree min-size max-size)
        (apply-color-by-partition)
        (apply-labels label-size false)
        (apply-layout simulation-time repulsion-strength)
        (if out-graph
          (export-to-pdf out-graph)
          (export-to-pdf (str dot-file ".pdf")))))
      (gen-stats)))


resources/jar-graph.sh
#!/bin/bash

JAR_EXEC=./bin/jar-graph-0.1.0-standalone.jar

if ! [ -x "$(command -v java)" ]; then
  echo 'Error: java is not installed.' >&2
  exit 1
elif ! [ -x "$(command -v jdeps)" ]; then
  echo 'Error: jdeps is not installed.' >&2
  exit 1
fi

usage()
{
  scriptname=$0
  echo "usage: $scriptname -f <jar_file> -p <package> [-t time] [-nh] [-o out_file] [-w dir] [-r repulsion] [-s font] [-m min] [-x max]"
  echo "  -t  time      specify simulation time in seconds"
  echo "  -n            no graph will be generated"
  echo "  -h            display help"
  echo "  -o  outfile   output file outfile"
  echo "  -w  workdir   working directory"
  echo "  -r  repulsion node repulsion strength value"
  echo "  -s  size      font size"
  echo "  -m  min       minimum node size"
  echo "  -x  max       maximum node size"
}

run_jdeps()
{
  if [ -z "$3" ]; then
    jdeps -verbose:class -dotoutput $1 $2;
  else
    package=$(sed "s/\./\\\./g" <<<"$3")
    jdeps -f '^(?!'$package').+' -verbose:class -dotoutput $1 $2;
  fi
}

prepare_graph()
{
  if [[ "$OSTYPE" == "darwin"* ]]; then
    sed -i "" "/(not found)/d" $2;
    sed -i "" "s/ (${1})//g" $2;
  else
    sed -i "/(not found)/d" $2;
    sed -i "" "s/ (${1})//g" $2;
  fi
}

cleanup_workdir()
{
  rm -f $1
  rm -f $2
}

target_jar=
package=
simulation_time_sec=20
no_graph=
out_file=
work_dir=/tmp/jar-graph
node_repulsion_strength=800
label_font_size=2
node_min_size=6
node_max_size=40

if [ $# -eq 0 ]; then
    usage
    exit 1
fi

while [ "$1" != "" ]; do
  case $1 in
    -f | --file )       shift
                        target_jar=$1
                        ;;
    -p | --package )    shift
                        package=$1
                        ;;
    -t | --time )       shift
                        simulation_time_sec=$1
                        ;;
    -n | --no-graph )   shift
                        no_graph=1
                        ;;
    -h | --help )       usage
                        exit
                        ;;
    -o | --out )        shift
                        out_file=$1
                        ;;
    -w | --workdir )    shift
                        work_dir=$1
                        ;;
    -r | --repulsion )  shift
                        node_repulsion_strength=$1
                        ;;
    -s | --font-size )  shift
                        label_font_size=$1
                        ;;
    -m | --min-node )   shift
                        node_min_size=$1
                        ;;
    -x | --max-node )   shift
                        node_max_size=$1
                        ;;
    * )                 usage
                        exit
                        ;;
  esac
  shift
done

if [ ! -f $target_jar ]; then
    echo "File not found!"
    exit 1
fi

mkdir -p $work_dir

name=$(basename $target_jar)
dotfile="$work_dir/${name}.dot"
out_file=$(pwd)/${out_file:-$name".pdf"}

run_jdeps $work_dir $target_jar $package 
prepare_graph $name $dotfile

java -jar $JAR_EXEC\
 --dot-file $dotfile\
 --out-graph $out_file\
 --simulation-time $simulation_time_sec\
 --repulsion-strength $node_repulsion_strength\
 --min-size $node_min_size\
 --max-size $node_max_size\
 --label-size $label_font_size\
 ${no_graph:+"--skip-graph"}

cleanup_workdir $dotfile "$work_dir/summary.dot"
